src/components/templates/ProtectedRoute.jsx

```diff
@@ -0,0 +1,51 @@
+import React, { useEffect, useState } from 'react';
+import { useNavigate } from 'react-router-dom';
+import Swal from 'sweetalert2';
+import routes from '../../constant/routes';
+import { loginNeedMessage, authInvalidMessage } from '../../utils/alert';
+
+const ProtectedRoute = ({ children, requiredAuth }) => {
+  const navigate = useNavigate();
+  const [isLoggedIn, setIsLoggedIn] = useState(false);
+  const [isAuthVaild, setIsAuthValid] = useState(false);
+
+  // eslint-disable-next-line consistent-return
+  const checkUserToken = () => {
+    const userToken = localStorage.getItem('accessToken');
+    if (!userToken || userToken === 'undefined') {
+      setIsLoggedIn(false);
+      Swal.fire(loginNeedMessage);
+      navigate(routes.login);
+    }
+    setIsLoggedIn(true);
+  };
+
+  const checkUserAuth = () => {
+    // requiredAuth(admin, student, user)가 존재할 경우, 현재 유저 권한에 대한 유효성을 검증
+    // 사실상 user 권한의 경우는 로그인 토큰 자체만으로도 검증이 가능하기 때문에, 별도로 requiredAuth prop을 넘겨줄 필요 없음
+    if (requiredAuth) {
+      const userAuth = localStorage.getItem('userAuth');
+      // [admin, student] 이렇게 넘겨줄 수도 있는 경우를 감안하여 다음과 같이 처리
+      // requiredAuth 중 하나라도 만족하면, 페이지 접근 권한 유효함
+      if (requiredAuth.includes(userAuth)) {
+        setIsAuthValid(true);
+      } else {
+        setIsAuthValid(false);
+        Swal.fire(authInvalidMessage);
+        navigate(routes.mypage);
+      }
+    } else {
+      // requiredAuth(페이지별 접근 권한)가 제공되지 않았을 때는 항상 권한이 유효하다고 간주합니다.
+      setIsAuthValid(true);
```

Comment:
```suggestion
    const userAuth = localStorage.getItem('userAuth');
    if (!requiredAuth || requiredAuth.includes(userAuth)) {  
      setIsAuthValid(true);
    }
    else {
        setIsAuthValid(false);
        Swal.fire(authInvalidMessage);
        navigate(routes.mypage);
      }
```

제가 코드를 완전히 이해한 것은 아니라 틀릴 수도 있습니다!!!
중첩 if문을 가만히 보다가 어떻게 줄일 수 있을까 고민해봤는데, 

1. `requiredAuth`가 false일때
2. `requiredAuth.includes(userAuth)`가 true일 때,

이 두가지 경우가 동일하게 `setIsAuthValid(true)` 로직만 있는 것 같더라구요..!
그래서 제가 한번 줄여봤는데..!!! 동작할지 안할지는 모르겠네욥..

Reply:
워후!! 같은 로직을 날카롭게 보시고 이렇게 깔쌈하게 줄여 주시다니....감사합니다!!👍👍
제 생각에는 배경님 말씀대로 바꿔도 로직자체에는 변화가 없어서 잘 작동할 것 같다고 생각했는데,
실제로 바꿔봐도 아주 잘되네요!! ㅎㅎ 조언해 주신 내용으로 변경해서 push 하겠습니다😄😄